src/common/base/CuckooHashCache.h

```diff
@@ -0,0 +1,875 @@
+        if (slot != -1) {
+            return table_position{i2, static_cast<size_t>(slot), ok};
+        }
+        return table_position{0, 0, failure_key_not_found};
+    }
+
+    // try_read_from_bucket will search the bucket for the given key and return
+    // the index of the slot if found, or -1 if not found.
+    template <typename K>
+    int try_read_from_bucket(const size_t bucket_ind, const partial_t partial, const K &key) {
+        bucket &b = buckets_[bucket_ind];
+        for (int i = 0; i < static_cast<int>(slot_per_bucket()); ++i) {
+            if (try_reclaim(bucket_ind, i)) {
+                continue;
+            }
+            if (!b.occupied(i) || (!is_simple() && partial != b.partial(i))) {
+                continue;
+            } else if (key_eq()(b.key(i), key)) {
+                ++locks_[lock_ind(bucket_ind)].hit_counter();
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    bool try_reclaim(const size_t bucket_ind, const size_t slot) {
+        bucket &b = buckets_[bucket_ind];
+        if (b.occupied(slot) &&
+            b.expire(slot) < nebula::time::WallClock::fastNowInSec()) {
+            del_from_bucket(bucket_ind, slot);
+            return true;
+        }
+        return false;
+    }
+
+    // cuckoo_insert tries to find an empty slot in either of the buckets to
+    // insert the given key into. It expects the locks to be taken outside the function. Before inserting, it
+    // checks that the key isn't already in the table. cuckoo hashing presents
+    // multiple concurrency issues, which are explained in the function. The
+    // following return states are possible:
+    //
+    // ok -- Found an empty slot, locks will be held on both buckets after the
+    // function ends, and the position of the empty slot is returned
+    //
+    // failure_key_duplicated -- Found a duplicate key, locks will be held, and
+    // the position of the duplicate key will be returned
+    //
+    // failure_table_full -- Failed to find an empty slot for the table. Locks
+    // are released. No meaningful position is returned.
+    template <typename K>
+    table_position cuckoo_insert(const hash_value hv, TwoBuckets &b, K &key) {
+        int res1, res2;
+        if (!try_find_insert_bucket(b.i1, res1, hv.partial, key)) {
+            return table_position{b.i1, static_cast<size_t>(res1),
+                                  failure_key_duplicated};
+        }
+        if (!try_find_insert_bucket(b.i2, res2, hv.partial, key)) {
+            return table_position{b.i2, static_cast<size_t>(res2),
+                                  failure_key_duplicated};
+        }
+        if (res1 != -1) {
+            return table_position{b.i1, static_cast<size_t>(res1), ok};
+        }
+        if (res2 != -1) {
+            return table_position{b.i2, static_cast<size_t>(res2), ok};
+        }
+        return table_position{0, 0, failure_table_full};
```

Comment:
If two slots are occupied, why not move the element on the original slot instead of return full directly. 

In some rare cases, there may be lots of empty slots, but return full just due to one key's  two slots are occupied?